From 39984ee6773d2f2d097241d5b5abfde474dbc16c Mon Sep 17 00:00:00 2001
From: Thomas Parrott <thomas.parrott@canonical.com>
Date: Mon, 9 Jan 2023 09:22:38 +0000
Subject: [PATCH] lxd/instance/drivers: Fixes delete of ephemeral VM on stop

Moves the IsRunning check out of the internal `delete()` function and into the
exported `Delete()` function. This avoids failing the the `delete()` call from
the `onStop()` function when trying to delete an ephemeral VM on stop.

Although containers were not affected (because their notion of running doesn't
include an ongoing Stop operation that VMs do) I've also moved the check in
the LXC driver to the same place for consistency.

Fixes #11261

Signed-off-by: Thomas Parrott <thomas.parrott@canonical.com>
---
 lxd/instance/drivers/driver_lxc.go  | 8 ++++----
 lxd/instance/drivers/driver_qemu.go | 8 ++++----
 2 files changed, 8 insertions(+), 8 deletions(-)

diff --git a/lxd/instance/drivers/driver_lxc.go b/lxd/instance/drivers/driver_lxc.go
index ab0b932186e4..0810f7f241e2 100644
--- a/lxd/instance/drivers/driver_lxc.go
+++ b/lxd/instance/drivers/driver_lxc.go
@@ -3600,15 +3600,15 @@ func (d *lxc) Delete(force bool) error {
 
 	defer op.Done(nil)
 
+	if d.IsRunning() {
+		return api.StatusErrorf(http.StatusBadRequest, "Instance is running")
+	}
+
 	return d.delete(force)
 }
 
 // Delete deletes the instance without creating an operation lock.
 func (d *lxc) delete(force bool) error {
-	if d.IsRunning() {
-		return api.StatusErrorf(http.StatusBadRequest, "Instance is running")
-	}
-
 	ctxMap := logger.Ctx{
 		"created":   d.creationDate,
 		"ephemeral": d.ephemeral,
diff --git a/lxd/instance/drivers/driver_qemu.go b/lxd/instance/drivers/driver_qemu.go
index ff238fb62cbe..8f7caaebe435 100644
--- a/lxd/instance/drivers/driver_qemu.go
+++ b/lxd/instance/drivers/driver_qemu.go
@@ -5090,15 +5090,15 @@ func (d *qemu) Delete(force bool) error {
 
 	defer op.Done(nil)
 
+	if d.IsRunning() {
+		return api.StatusErrorf(http.StatusBadRequest, "Instance is running")
+	}
+
 	return d.delete(force)
 }
 
 // Delete the instance without creating an operation lock.
 func (d *qemu) delete(force bool) error {
-	if d.IsRunning() {
-		return api.StatusErrorf(http.StatusBadRequest, "Instance is running")
-	}
-
 	ctxMap := logger.Ctx{
 		"created":   d.creationDate,
 		"ephemeral": d.ephemeral,
